group-membership/src/Code.js

```diff
@@ -0,0 +1,150 @@
+  requiredFieldMissing: 'Required field(s) missing: fill out all fields for this row',
+  allowedFieldNotSpecified: '',
+};
+
+/**
+ * Installs a trigger in the Spreadsheet to run upon the Sheet being opened.
+ * To learn more about triggers read:
+ * https://developers.google.com/apps-script/guides/triggers
+ */
+function onOpen() {
+  SpreadsheetApp.getUi()
+      .createMenu('Install trigger')
+      .addItem('onEdit', 'installOnEditTrigger')
+      .addToUi();
+}
+
+/**
+ * Installs a trigger in the Spreadsheet that is scheduled
+ * to run upon when values in the Sheet are edited.
+ * To learn more about triggers read:
+ * https://developers.google.com/apps-script/guides/triggers/installable
+ */
+function installOnEditTrigger() {
+  ScriptApp.newTrigger('onEditInstallableTrigger')
+      .forSpreadsheet(SpreadsheetApp.getActive())
+      .onEdit()
+      .create();
+}
+
+/**
+ * Trigger that runs on edit after being installed via the interface.
+ *
+ * @param {Object} e - onEdit trigger event.
+ */
+function onEditInstallableTrigger(e) {
+  // Get the headers, row range and values from the active sheet.
+  const sheet = SpreadsheetApp.getActiveSheet();
+  const headers = sheet.getDataRange().offset(0, 0, 1).getValues()[0];
+  const range = sheet.getRange(e.range.getRow(), 1, 1, headers.length);
+  const row = range.getValues()[0];
+
+  // Convert the row Array into an entries Object using the headers for the
+  // field names.
+  const entries = headers.reduce((result, columnName, i) => {
+    result[columnName] = row[i];
+    return result;
+  }, {});
+
+  // Update the entries Object with the email state returned by addToGroup().
+  try {
+    const emailState = addToGroup(
+        entries[EMAIL],
+        entries[GOOGLE_GROUP],
+        entries[ALLOWED],
+        entries[EMAIL_TEMPLATE_DOC_URL],
+        entries[EMAIL_SUBJECT]
+    );
+    entries[EMAIL_STATE] = emailState == EMAIL_STATE_VALUE.sent ?
+        `${emailState}: ${new Date()}` : emailState;
+  } catch (e) {
+    // If there's an error, report that as the email state.
+    entries[EMAIL_STATE] = e;
+  }
+
+  // Convert the updated entries Object into a row Array.
+  const rowToWrite = headers.map((columnName) => entries[columnName]);
+
+  // setValues() receives a 2D array, so we create an array with the row
+  // contents.
+  range.setValues([rowToWrite]);
+}
+
+
+/**
+ * Trigger that runs on edit after being installed via the interface.
+ *
+ * @param {string} userEmail - email of user to add to the group.
+ * @param {string} groupEmail - address of Google Group to add user to.
+ * @param {string} allowed - 'yes' flag to add user to group.
+ * @param {string} emailTemplateDocUrl - Google Doc URL that serves as template
+ *        of the welcome email sent to a user added to the group.
+ * @param {string} emailSubject - subject of welcome email sent to user added
+ *        to group.
+ * @return {string} - state if email was sent to a user added in the sheet.
```

Comment:
I like that you're returning the error to the user in the Spreadsheet but could you change the return type of the method and log the string? It is harder to later debug string return values.

Reply:
Makes the sample very long now, but these changes have been implemented per your request.